
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>geppy.tools package &#8212; geppy 0.1.0a0 documentation</title>
    <link rel="stylesheet" href="_static/pydoctheme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
 
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-15144370-3']);
  _gaq.push(['_trackPageview']);
</script>
<script type="text/javascript" src="_static/copybutton.js"></script>

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
    <li><a href="https://github.com/ShuhuaGao/geppy"><img src="_static/geppy-icon.png" style="height: 30px; vertical-align: middle; padding-right: 1em;" />Project Homepage</a> &#187;</li>
	<li><a href="index.html">geppy 0.1.0a0 documentation</a> &#187; </li>
 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="geppy-tools-package">
<h1>geppy.tools package<a class="headerlink" href="#geppy-tools-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-geppy.tools.crossover">
<span id="geppy-tools-crossover-module"></span><h2>geppy.tools.crossover module<a class="headerlink" href="#module-geppy.tools.crossover" title="Permalink to this headline">¶</a></h2>
<p>The module <code class="xref py py-mod docutils literal notranslate"><span class="pre">crossover</span></code> provides crossover (mating, recombination) related genetic modifications in GEP,
including one-point and two-point crossover, and gene crossover between multigenic chromosomes.
Please refer to Chapter 3 of <a class="reference internal" href="#fc2006" id="id1">[FC2006]</a> for more details.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">All the recombination operators including <a class="reference internal" href="#geppy.tools.crossover.crossover_one_point" title="geppy.tools.crossover.crossover_one_point"><code class="xref py py-func docutils literal notranslate"><span class="pre">crossover_one_point()</span></code></a>, <a class="reference internal" href="#geppy.tools.crossover.crossover_two_point" title="geppy.tools.crossover.crossover_two_point"><code class="xref py py-func docutils literal notranslate"><span class="pre">crossover_two_point()</span></code></a>, and
<a class="reference internal" href="#geppy.tools.crossover.crossover_gene" title="geppy.tools.crossover.crossover_gene"><code class="xref py py-func docutils literal notranslate"><span class="pre">crossover_gene()</span></code></a> can be applied to both <a class="reference internal" href="geppy.core.html#geppy.core.entity.Gene" title="geppy.core.entity.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a>
and <a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>.</p>
</div>
<table class="docutils citation" frame="void" id="fc2006" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[FC2006]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id2">2</a>, <a class="fn-backref" href="#id3">3</a>, <a class="fn-backref" href="#id4">4</a>)</em> Ferreira, Cândida. Gene expression programming: mathematical modeling by an artificial
intelligence. Vol. 21. Springer, 2006.</td></tr>
</tbody>
</table>
<dl class="function">
<dt id="geppy.tools.crossover.crossover_gene">
<code class="descclassname">geppy.tools.crossover.</code><code class="descname">crossover_gene</code><span class="sig-paren">(</span><em>ind1</em>, <em>ind2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/crossover.html#crossover_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.crossover.crossover_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Entire genes are exchanged between two parent chromosomes. The two individuals are modified in place, and the two
children are returned.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ind1</strong> – The first individual (chromosome) participating in the crossover.</li>
<li><strong>ind2</strong> – The second individual (chromosome) participating in the crossover.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple of two children individuals</p>
</td>
</tr>
</tbody>
</table>
<p>For multigenic chromosomes, only a single gene index is randomly chosen and the two genes at that position are
swapped. This operation has no effect if the chromosome has only one gene. Typically, a gene recombination rate
around 0.3 is used.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.crossover.crossover_one_point">
<code class="descclassname">geppy.tools.crossover.</code><code class="descname">crossover_one_point</code><span class="sig-paren">(</span><em>ind1</em>, <em>ind2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/crossover.html#crossover_one_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.crossover.crossover_one_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute one-point recombination of two individuals. The two individuals are modified in place, and the two children
are returned.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ind1</strong> – The first individual (chromosome) participating in the crossover.</li>
<li><strong>ind2</strong> – The second individual (chromosome) participating in the crossover.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A tuple of two children individuals.</p>
</td>
</tr>
</tbody>
</table>
<p>Note the crossover can happen at any point across the whole chromosome and thus entire genes may be also exchanged
between the two parents if they are multigenic chromosomes.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.crossover.crossover_two_point">
<code class="descclassname">geppy.tools.crossover.</code><code class="descname">crossover_two_point</code><span class="sig-paren">(</span><em>ind1</em>, <em>ind2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/crossover.html#crossover_two_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.crossover.crossover_two_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute two-point recombination of two individuals. The two individuals are modified in place, and the two children
are returned. The materials between two randomly chosen points are swapped to generate two children.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ind1</strong> – The first individual (chromosome) participating in the crossover.</li>
<li><strong>ind2</strong> – The second individual (chromosome) participating in the crossover.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A tuple of two individuals.</p>
</td>
</tr>
</tbody>
</table>
<p>Note the crossover can happen at any point across the whole chromosome and thus entire genes may be also exchanged
between the two parents if they are multigenic chromosomes.</p>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.generator">
<span id="geppy-tools-generator-module"></span><h2>geppy.tools.generator module<a class="headerlink" href="#module-geppy.tools.generator" title="Permalink to this headline">¶</a></h2>
<p>This module <code class="xref py py-mod docutils literal notranslate"><span class="pre">generator</span></code> provides functionality to generate a genome for a gene. That is, choose functions and
terminals randomly from a given primitive set to form a linear form gene expression.</p>
<dl class="function">
<dt id="geppy.tools.generator.generate_dc">
<code class="descclassname">geppy.tools.generator.</code><code class="descname">generate_dc</code><span class="sig-paren">(</span><em>rnc_array_length</em>, <em>dc_length</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/generator.html#generate_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.generator.generate_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Dc domain for a RNC array of size <em>rnc_array_length</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>rnc_array_length</strong> – length of the RNC array</li>
<li><strong>dc_length</strong> – length of the Dc domain</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a list of integers of length <em>dc_length</em>, each element in range <cite>[0, rnc_array_length - 1]</cite></p>
</td>
</tr>
</tbody>
</table>
<p>Refer to Chapter 5 of <a class="reference internal" href="#fc2006" id="id2">[FC2006]</a> for more details.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.generator.generate_genome">
<code class="descclassname">geppy.tools.generator.</code><code class="descname">generate_genome</code><span class="sig-paren">(</span><em>pset</em>, <em>head_length</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/generator.html#generate_genome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.generator.generate_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a genome with the given primitive set <em>pset</em> and the specified head domain length <em>head_length</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pset</strong> – a primitive set</li>
<li><strong>head_length</strong> – length of the head domain</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a list of symbols representing a genome</p>
</td>
</tr>
</tbody>
</table>
<p>Supposing the maximum arity of functions in <em>pset</em> is <em>max_arity</em>, then the tail length is automatically
determined to be <code class="docutils literal notranslate"><span class="pre">tail_length</span> <span class="pre">=</span> <span class="pre">head_length</span> <span class="pre">*</span> <span class="pre">(max_arity</span> <span class="pre">-</span> <span class="pre">1)</span> <span class="pre">+</span> <span class="pre">1</span></code>.</p>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.mutation">
<span id="geppy-tools-mutation-module"></span><h2>geppy.tools.mutation module<a class="headerlink" href="#module-geppy.tools.mutation" title="Permalink to this headline">¶</a></h2>
<p>The module <code class="xref py py-mod docutils literal notranslate"><span class="pre">mutation</span></code> provides mutation related genetic modification operators in GEP, including point-mutation,
transposition and inversion. Please refer to Chapter 3 of <a class="reference internal" href="#fc2006" id="id3">[FC2006]</a> for more details.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ol class="last arabic simple">
<li>Operators that can be applied to both <a class="reference internal" href="geppy.core.html#geppy.core.entity.Gene" title="geppy.core.entity.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a> and <a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>:
<a class="reference internal" href="#geppy.tools.mutation.mutate_uniform" title="geppy.tools.mutation.mutate_uniform"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate_uniform()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.invert" title="geppy.tools.mutation.invert"><code class="xref py py-func docutils literal notranslate"><span class="pre">invert()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.is_transpose" title="geppy.tools.mutation.is_transpose"><code class="xref py py-func docutils literal notranslate"><span class="pre">is_transpose()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.ris_transpose" title="geppy.tools.mutation.ris_transpose"><code class="xref py py-func docutils literal notranslate"><span class="pre">ris_transpose()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.gene_transpose" title="geppy.tools.mutation.gene_transpose"><code class="xref py py-func docutils literal notranslate"><span class="pre">gene_transpose()</span></code></a></li>
<li>Operators that end with ‘_dc’ are specially designed to handle the Dc domain in
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>.</li>
</ol>
</div>
<dl class="function">
<dt id="geppy.tools.mutation.gene_transpose">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">gene_transpose</code><span class="sig-paren">(</span><em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#gene_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.gene_transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform gene transposition in place</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a tuple of one individual</td>
</tr>
</tbody>
</table>
<p>An entire gene is selected randomly and exchanged with the first gene in the chromosome.
Obviously, this operation only makes sense if the chromosome is a multigenic one. Typically, the gene transposition
rate is set around 0.1.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.invert">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">invert</code><span class="sig-paren">(</span><em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#invert"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.invert" title="Permalink to this definition">¶</a></dt>
<dd><p>A gene is randomly chosen, and afterwards a subsequence within this gene’s head domain is randomly selected
and inverted.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a tuple of one individual</td>
</tr>
</tbody>
</table>
<p>Typically, a small inversion rate of 0.1 is used.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.invert_dc">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">invert_dc</code><span class="sig-paren">(</span><em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#invert_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.invert_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Dc-specific inversion. A gene is randomly chosen, and afterwards a subsequence within this gene’s Dc domain is
randomly selected and inverted.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a tuple of one individual</td>
</tr>
</tbody>
</table>
<p>Typically, a small inversion rate of 0.1 is used.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.is_transpose">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">is_transpose</code><span class="sig-paren">(</span><em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#is_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.is_transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform IS transposition in place</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a tuple of one individual</td>
</tr>
</tbody>
</table>
<p>An Insertion Sequence (IS) is a random chosen segment across the chromosome, and then the
IS is copied to be inserted at another position in the head of gene, except the start position. Note that an IS from
one gene may be copied to a different gene.
Typically, the IS transposition rate is set around 0.1.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.mutate_rnc_array_dc">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">mutate_rnc_array_dc</code><span class="sig-paren">(</span><em>individual</em>, <em>rnc_gen</em>, <em>indpb</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_rnc_array_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_rnc_array_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct mutation of RNCs, which changes the values in a gene’s RNC array
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc.rnc_array" title="geppy.core.entity.GeneDc.rnc_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rnc_array()</span></code></a> randomly.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>.</li>
<li><strong>rnc_gen</strong> – callable, which returns a random numerical constant by calling <code class="docutils literal notranslate"><span class="pre">rnc_gen()</span></code>.</li>
<li><strong>indpb</strong> – individual probability of a RNC being mutated</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple of one individual</p>
</td>
</tr>
</tbody>
</table>
<p>The genetic operators <a class="reference internal" href="#geppy.tools.mutation.mutate_uniform_dc" title="geppy.tools.mutation.mutate_uniform_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate_uniform_dc()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.transpose_dc" title="geppy.tools.mutation.transpose_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">transpose_dc()</span></code></a> and <a class="reference internal" href="#geppy.tools.mutation.invert_dc" title="geppy.tools.mutation.invert_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">invert_dc()</span></code></a> actually only move the
random numerical constants around without generating new numerical values. This method <a class="reference internal" href="#geppy.tools.mutation.mutate_rnc_array_dc" title="geppy.tools.mutation.mutate_rnc_array_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate_rnc_array_dc()</span></code></a>
can replace the value of a particular numerical constant by another in the
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc.rnc_array" title="geppy.core.entity.GeneDc.rnc_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rnc_array()</span></code></a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ol class="arabic simple">
<li>The direct mutation of random numerical constants might also have a great impact in the overall solution.</li>
<li>Interestingly, the direct mutation of RNCs seems to have a very limited impact on the creation of good
models and better results may be obtained when this operator is switched off.</li>
<li>Potentially, a well-dimensioned initial diversity of RNCs is more than sufficient to allow their evolutionary
tuning, as they are constantly being moved around by all the genetic operators.</li>
<li>Thus, if this operator is adopted, a very small <em>indpb</em> should be set, like 0.01.</li>
</ol>
<p class="last">Refer to section 5.4.4 of <a class="reference internal" href="#fc2006" id="id4">[FC2006]</a> for more details.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.mutate_uniform">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">mutate_uniform</code><span class="sig-paren">(</span><em>individual</em>, <em>pset</em>, <em>indpb</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_uniform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_uniform" title="Permalink to this definition">¶</a></dt>
<dd><p>Uniform point mutation. For each symbol in <em>individual</em>, change it to another randomly chosen symbol from <em>pset</em>
with the probability <em>indpb</em>. A symbol may be a function or a terminal.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, the chromosome to be mutated.</li>
<li><strong>pset</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">PrimitiveSet</span></code>, a primitive set</li>
<li><strong>indpb</strong> – probability of mutating each symbol</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A tuple of one chromosome</p>
</td>
</tr>
</tbody>
</table>
<p>It is typical to set a mutation rate <em>indpb</em> equivalent to two one-point mutations per chromosome. That is,
<code class="docutils literal notranslate"><span class="pre">indpb</span> <span class="pre">=</span> <span class="pre">2</span> <span class="pre">/</span> <span class="pre">len(chromosome)</span> <span class="pre">*</span> <span class="pre">len(gene)</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.mutate_uniform_dc">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">mutate_uniform_dc</code><span class="sig-paren">(</span><em>individual</em>, <em>indpb</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_uniform_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_uniform_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Dc-specific mutation. This operator changes one index stored in the Dc domain to another index in place. The indices
in the Dc domain can later be used to retrieve numerical constants from the gene associated
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc.rnc_array" title="geppy.core.entity.GeneDc.rnc_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rnc_array()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a></li>
<li><strong>indpb</strong> – probability of mutating each index/position in the Dc domain</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple of one chromosome</p>
</td>
</tr>
</tbody>
</table>
<p>It is typical to set a mutation rate <em>indpb</em> equivalent to two one-point mutations per chromosome. That is,
<code class="docutils literal notranslate"><span class="pre">indpb</span> <span class="pre">=</span> <span class="pre">2</span> <span class="pre">/</span> <span class="pre">len(chromosome)</span> <span class="pre">*</span> <span class="pre">len(gene)</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.ris_transpose">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">ris_transpose</code><span class="sig-paren">(</span><em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#ris_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.ris_transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform RIS transposition in place</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a tuple of one individual</td>
</tr>
</tbody>
</table>
<p>A Root Insertion Sequence (RIS) is a segment of consecutive elements that starts with a
function. Once an RIS is randomly chosen, it is copied and inserted at the root (first position) of a gene. Note
that an RIS from one gene may be copied to a different gene.
Typically, the RIS transposition rate is set around 0.1.</p>
</dd></dl>

<dl class="function">
<dt id="geppy.tools.mutation.transpose_dc">
<code class="descclassname">geppy.tools.mutation.</code><code class="descname">transpose_dc</code><span class="sig-paren">(</span><em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#transpose_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.transpose_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>This operator randomly chooses the chromosome, the gene with its respective Dc to be subjected to transposition,
the start and termination points of the transposon, and the target site; then it inserts a copy of the transposon
from the place of origin to the target site, while the old content at the target site are shifted to the right
and the length of the Dc domain is maintained.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a tuple of one individual</td>
</tr>
</tbody>
</table>
<p>Typically, a small transposition rate of 0.1 is used. Note that a transposon from one gene’s Dc may be copied and
inserted into another gene’s Dc, which means the genes in a multigenic chromosome <em>individual</em>
must have an identical structure.</p>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.parser">
<span id="geppy-tools-parser-module"></span><h2>geppy.tools.parser module<a class="headerlink" href="#module-geppy.tools.parser" title="Permalink to this headline">¶</a></h2>
<p>This module <a class="reference external" href="https://docs.python.org/3/library/parser.html#module-parser" title="(in Python v3.7)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">parser</span></code></a> provides functionality for compiling an individual (a chromosome) in GEP into an executable
lambda function in Python for subsequent fitness evaluation.</p>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">Parse an individual into codes of other languages, such as C++/Java, for deployment in an industrial environment.</p>
</div>
<dl class="function">
<dt id="geppy.tools.parser.compile_">
<code class="descclassname">geppy.tools.parser.</code><code class="descname">compile_</code><span class="sig-paren">(</span><em>individual</em>, <em>pset</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/parser.html#compile_"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.parser.compile_" title="Permalink to this definition">¶</a></dt>
<dd><p>Compile the individual into a Python lambda expression.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>individual</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code>, a chromosome</li>
<li><strong>pset</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">PrimitiveSet</span></code>, a primitive set</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a function if the primitive set <em>pset</em> has any inputs (arguments), which can later be called with
specific parameter values; otherwise, a numerical result obtained from evaluation.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-geppy.tools" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">geppy.tools package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-geppy.tools.crossover">geppy.tools.crossover module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.generator">geppy.tools.generator module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.mutation">geppy.tools.mutation module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.parser">geppy.tools.parser module</a></li>
<li><a class="reference internal" href="#module-geppy.tools">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/geppy.tools.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
    <li><a href="https://github.com/ShuhuaGao/geppy"><img src="_static/geppy-icon.png" style="height: 30px; vertical-align: middle; padding-right: 1em;" />Project Homepage</a> &#187;</li>
	<li><a href="index.html">geppy 0.1.0a0 documentation</a> &#187; </li>
 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2018, Shuhua Gao.
    <br />
    Built on None.
    <a href="bugs.html">Found a bug</a>?
    <br />
    Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.7.4.
</div>
<script type="text/javascript">
  (function() {
     var ga = document.createElement('script');
         ga.src = ('https:' == document.location.protocol ?
                'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    ga.setAttribute('async', 'true');
    document.documentElement.firstChild.appendChild(ga);
  })();
</script>

  </body>
</html>